controllers/remove_process_groups.go

```diff
@@ -77,7 +80,26 @@ func (u removeProcessGroups) reconcile(r *FoundationDBClusterReconciler, context
 		}
 	}
 
-	removedProcessGroups := r.removeProcessGroups(context, cluster, processGroupsToRemove)
+	status, err := adminClient.GetStatus()
+	if err != nil {
+		return &requeue{curError: err}
+	}
+
+	// In addition to that we should add the same logic as in the exclude step
+	// to ensure we never exclude/remove more process groups than desired.
+	zonedRemovals, err := removals.GetZonedRemovals(status, processGroupsToRemove)
+	if err != nil {
+		return &requeue{curError: err}
+	}
+
+	zone, zoneRemovals, err := removals.GetProcessGroupsToRemove(cluster.GetRemovalMode(), zonedRemovals)
+	if err != nil {
+		return &requeue{curError: err}
+	}
+
+	logger.Info("Removing process groups", "zone", zone, "count", len(zoneRemovals), "deletionMode", string(cluster.Spec.AutomationOptions.DeletionMode))
+
+	removedProcessGroups := r.removeProcessGroups(context, cluster, zoneRemovals)
```

Comment:
In theory we should requeue here if we have some zones left. In practice we will requeue anyway because of the generation change in the Pod, PVC, service.

I will add a test case for resources that hang in a terminating state to ensure we exclude those.

Reply:
I think we can have a reasonable heuristic based around the deletionTimestamp on the pod. If it's in the last 10 minutes or so, we can treat it as an active deletion and wait before deleting any further resources. If it's older than that, it's likely a stuck termination and we can ignore it and continue deleting further resources.